### PR TITLE
OPENEUROPA-2619: Create Project corporate content type

### PR DESCRIPTION
**Pull request by drishu**
### Description

Create Project corporate content type: module scaffold, fields export, behat test.

### Change log

- Added:
  - project content type and fields
  - shared field oe_reference
- Changed:

### CODE REVIEW

File: tests/Behat/Content/Node/ProjectContentContext.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?php
```
Comment (drishu) [inline tests/Behat/Content/Node/ProjectContentContext.php:1]: can you test if we need this file ? try removing and run the behat tests, if they are green, then commit the removal

Comment (julien-) [reply tests/Behat/Content/Node/ProjectContentContext.php]: Yeah it works. Changed doing `git commit --amend`

I don't really understand the existence of tests/Behat/Content/Node/**Event**ContentContext.php

File: oe_content.install [hunk 2]
```diff
@@ -37,3 +39,17 @@ function oe_content_update_8001() {
     $entity_definition_update_manager->updateFieldStorageDefinition($field_definition);
   }
 }
+
+/**
+ * Update field storage with oe_reference.
+ */
+function oe_content_update_8002() {
```
Comment (ademarco) [inline oe_content.install:46]: We stopped using update hooks as you might not have access to the full Drupal API here. Move this in `oe_content.post_update.php` and rename it following the convention there.

File: oe_content.install [hunk 3]
```diff
@@ -37,3 +39,17 @@ function oe_content_update_8001() {
...
+  // Obtain configuration from yaml files.
+  $storage = new FileStorage(drupal_get_path('module', 'oe_content') . '/config/install');
+
+  // Obtain the storage manager for field storage bases,
+  // Create a new field from the yaml configuration and save.
+  \Drupal::entityManager()->getStorage('field_storage_config')
+    ->create($storage->read('field.storage.node.oe_reference'))
+    ->save();
```
Comment (ademarco) [inline oe_content.install:54]: Let's use Drupal API to create the base field:

```php
use Drupal\field\Entity\FieldStorageConfig;

<?php
    FieldStorageConfig::create([
      'field_name' => 'oe_reference',
      'entity_type' => 'node',
      ....
    ])->save();
```

File: modules/oe_content_project/config/install/field.field.node.oe_project.oe_author.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  module:
+    - rdf_skos
+id: node.oe_project.oe_author
+field_name: oe_author
+entity_type: node
+bundle: oe_project
+label: Author
+description: ''
+required: true
+translatable: true
```
[the remark this answers is not in the document]
Comment (julien-) [reply modules/oe_content_project/config/install/field.field.node.oe_project.oe_author.yml:16]: Fixed

File: modules/oe_content_project/config/install/field.field.node.oe_project.oe_subject.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  module:
+    - rdf_skos
+id: node.oe_project.oe_subject
+field_name: oe_subject
+entity_type: node
+bundle: oe_project
+label: Subject
+description: 'The topics mentioned on this page. These will be used by search engines and dynamic lists to determine their relevance to a user.'
+required: true
+translatable: true
```
[the remark this answers is not in the document]
Comment (julien-) [reply modules/oe_content_project/config/install/field.field.node.oe_project.oe_subject.yml:16]: Fixed

File: modules/oe_content_project/oe_content_project.module [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+/**
+ * @file
+ * The OpenEuropa Content Project module.
+ */
+
+declare(strict_types = 1);
```
[the remark this answers is not in the document]
Comment (julien-) [reply modules/oe_content_project/oe_content_project.module:8]: Fixed

File: tests/features/project.feature [hunk 7]
```diff
@@ -0,0 +1,43 @@
+@api @project
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/features/project.feature:1]: Fixed

File: tests/features/project.feature [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+Feature: Project content creation
+  In order to have projects on the site
+  As an editor
+  I need to be able to create and see project items
+
+  @javascript @av_portal
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/features/project.feature:7]: Fixed

File: tests/Behat/DateFieldContext.php [hunk 9]
```diff
@@ -111,6 +111,9 @@ public function fillDateRangeSelectListField(string $field_item, string $field_g
       'Minute' => 'i',
     ];
     foreach ($date_components as $date_component => $date_component_format) {
+      if (!$field_selector->hasSelect($date_component)) {
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/Behat/DateFieldContext.php:116]: This function receive datetime text input like "_23-02-2019 14:15_" and set each element list date format (One HTML <select> for one element like : Year, Month, Day..)
Sometimes as in the actual case, we don't have Hour and Minutes so the `if` is to manage cases that the <select> of Hour and Minute does not exist.

File: tests/features/content-types.feature [hunk 10]
```diff
@@ -5,7 +5,7 @@ Feature: Common fields across the content types
   I should have common fields configured the same for each content type.
 
   Scenario Outline: Fields have the correct description.
-    Given I am logged in as a user with the "create oe_event content, create oe_news content, create oe_page content, create oe_policy content, create oe_publication content, access content" permission
+    Given I am logged in as a user with the "create oe_event content, create oe_news content, create oe_page content, create oe_policy content, create oe_project content, create oe_publication content, access content" permission
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/features/content-types.feature:8]: It depends if we have to set same field description for the content type Project than the others or not.
Nothing is defined for this in the ticket

[the remark this answers is not in the document]
Comment (ademarco) [reply tests/features/content-types.feature:8]: Yeah, those description should be added on all content types.

File: tests/Behat/DateFieldContext.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (ademarco) [reply tests/Behat/DateFieldContext.php:116]: Ok cool, please add a comment on top of the `if` describing this case, thanks!

File: tests/features/project.feature [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    When I visit "the Project creation page"
+    And I fill in "Title" with "My Project"
+    And I fill in "Subject" with "EU financing"
+    And I fill in "Content owner" with "Committee on Agriculture and Rural Development"
+    And I fill in "Author" with "European Patent Office"
+    And I fill in "Departments" with "Audit Board of the European Communities"
+    And I fill in "Teaser" with "Project teaser"
+    And I fill in "Summary" with "Summary text"
+    And I fill in "Body text" with "Body text"
```
[the remark this answers is not in the document]
Comment (julien-) [reply tests/features/project.feature:20]: Fixed

File: tests/Behat/DateFieldContext.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (julien-) [reply tests/Behat/DateFieldContext.php:117]: Comment implemented

File: tests/features/content-types.feature (hunk 10 above)
[the remark this answers is not in the document]
Comment (julien-) [reply tests/features/content-types.feature:8]: Fixed

File: oe_content.install (hunk 3 above)
Comment (julien-) [reply oe_content.install]: Fixed

File: oe_content.install (hunk 2 above)
Comment (julien-) [reply oe_content.install]: Fixed

File: tests/Behat/DateFieldContext.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (ademarco) [reply tests/Behat/DateFieldContext.php:117]: You comment here above was much more descriptive that that one put in the code! :smile: could you please use that one instead? The one added in the code is cryptic.

File: tests/features/project.feature [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    And I fill in "URL" with "http://example.com" in the "Call for proposals" region
+    And I fill in "Link text" with "Example proposal" in the "Call for proposals" region
+    And I fill in "Results" with "Result 1 text" in the "Result" region
+    And I fill in "Result files" with "My Document 1" in the "Result" region
+    And I fill in "Alternative title" with "My alternative title text" in the "Alternative titles and teaser" region
+    And I fill in "Navigation title" with "My navigation title text" in the "Alternative titles and teaser" region
+    And I fill in "Departments" with "Audit Board of the European Communities"
+
+    # Add link fields.
+
```
Comment (ademarco) [inline tests/features/project.feature:42]: Remove empty line.

File: tests/features/project.feature [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    When I press "Save"
+
```
Comment (ademarco) [inline tests/features/project.feature:44]: Remove empty line.

File: tests/features/project.feature (hunk 12 above)
Comment (julien-) [reply tests/features/project.feature]: Fixed

File: tests/features/project.feature (hunk 13 above)
Comment (julien-) [reply tests/features/project.feature]: Fixed

File: tests/Behat/DateFieldContext.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (julien-) [reply tests/Behat/DateFieldContext.php:122]: Fixed ... hoping it was your idea

File: modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_dates.yml [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    - field.storage.node.oe_project_dates
+    - node.type.oe_project
+  module:
+    - datetime_range
+id: node.oe_project.oe_project_dates
+field_name: oe_project_dates
+entity_type: node
+bundle: oe_project
+label: 'Project period'
+description: 'If project has no end date leave it empty or use the same start date value'
```
Comment (nagyad) [inline modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_dates.yml:14]: image
I tried to leave the end date empty, I can't save the form due to validation error: "This value should not be null."

File: modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_budget.yml [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+required: false
+translatable: false
+default_value: {  }
+default_value_callback: ''
+settings:
+  min: null
+  max: null
+  prefix: €
+  suffix: ''
+field_type: float
```
Comment (nagyad) [inline modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_budget.yml:22]: I know that mostly budgets are round numbers, but I did not find any requirements in the specification regarding this, so business should know about the fact that in the float field their 90009 values will be rounded up. If that's correct, please ignore my comment.

File: modules/oe_content_project/config/install/core.entity_form_display.node.oe_project.default.yml [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+    - datetime_range
+    - entity_reference_revisions
+    - field_group
+    - link
+    - path
+    - rdf_skos
+    - text
+third_party_settings:
+  field_group:
+    group_result:
```
Comment (nagyad) [inline modules/oe_content_project/config/install/core.entity_form_display.node.oe_project.default.yml:30]: Does a result file make sense without the result field filled in? I could not find anything related to this in the COMM specifications.

File: oe_content.post_update.php [hunk 17]
```diff
@@ -129,3 +130,17 @@ function oe_content_post_update_00004(): void {
...
+/**
+ * Update field storage with oe_reference.
+ */
+function oe_content_post_update_00005(): void {
+  // Obtain configuration from yaml files.
+  $storage = new FileStorage(drupal_get_path('module', 'oe_content') . '/config/post_updates/00005_add_oe_reference_storage_field');
+
+  // Obtain the storage manager for field storage bases,
+  // Create a new field from the yaml configuration and save.
+  \Drupal::entityManager()->getStorage('field_storage_config')
```
[the remark this answers is not in the document]
Comment (julien-) [reply oe_content.post_update.php:143]: fixed

File: modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_budget.yml (hunk 15 above)
Comment (ademarco) [reply modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_budget.yml:22]: This is correctly set to float, as per COMM requirements.

File: modules/oe_content_project/config/install/core.entity_form_display.node.oe_project.default.yml (hunk 16 above)
Comment (ademarco) [reply modules/oe_content_project/config/install/core.entity_form_display.node.oe_project.default.yml:30]: I can't find anything of the sort in the requirements, we can leave it like this and, if necessary, we will add extra validation later on.

File: modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_dates.yml (hunk 14 above)
Comment (ademarco) [reply modules/oe_content_project/config/install/field.field.node.oe_project.oe_project_dates.yml]: After discussing internally we have decided to change the description to adapt to Drupal's behavior. We need to check if it makes sense to not have an and date for projects at all.

**Pull request merged**